Copter: guided mode update to use PosControl tools

lthall opened the pull request:
This PR converts Guided Mode to use the new Position Controller.

It also adds the ability to command accelerations.

rmackay9 commented on ArduCopter/GCS_Mavlink.cpp:
```diff
@@ -143,39 +144,38 @@ void GCS_MAVLINK_Copter::send_position_target_local_ned()
-        target_vel = copter.flightmode->get_vel_desired_cms() * 0.01f;
         break;
     }
+    target_pos = copter.mode_guided.get_target_pos() * 0.01f; // convert to metres
```
this leads to the possibility that we could provide non-zero targets in the message but the type_mask could say they should be avoided.

So for example, we could be flying to a position but if copter.mode_guided.get_target_vel() returned a non-zero or out-of-date value it would appear in the message.

If we think that this change is ok then we should at least move the declaration of the vectors down here.  eg, "const Vector3f target_pos = copter.mode_guided.get_target_pos() * 0.01f; // convert to metres"

rmackay9 commented on ArduCopter/GCS_Mavlink.cpp:
```diff
@@ -143,39 +144,38 @@ void GCS_MAVLINK_Copter::send_position_target_local_ned()
+        target_accel.x, // afx
+        target_accel.y, // afy
+        -target_accel.z,// afz
+        copter.pos_control->get_yaw_cd() * 0.01f, // yaw
```
we are providing yaw and yaw rate values but the type_mask says they should always be ignored which seems odd.

rmackay9 commented on ArduCopter/GCS_Mavlink.cpp:
```diff
@@ -1172,11 +1183,11 @@ void GCS_MAVLINK_Copter::handleMessage(const mavlink_message_t &msg)
-        } else if (!pos_ignore && vel_ignore) {
-            copter.mode_guided.set_destination(pos_vector, !yaw_ignore, yaw_cd, !yaw_rate_ignore, yaw_rate_cds, yaw_relative);
+            copter.mode_guided.set_velaccel(vel_vector, accel_vector, !yaw_ignore, yaw_cd, !yaw_rate_ignore, yaw_rate_cds, yaw_relative);
+        } else if (!pos_ignore && vel_ignore && acc_ignore) {
```
I guess this means that we will throw away requests that have position and acceleration but no velocity.  E.g. !pos_ignore && vel_ignore && !acc_ignore.  I guess we can't just set set_destination_posvelaccel() with a zero velocity?

lthall replied:
I might be able to make this work but for now it seems too much of a corner case. We can focus on the cases that make immediate sense for now.

rmackay9 commented on ArduCopter/mode.h:
```diff
@@ -845,11 +845,20 @@ class ModeGuided : public Mode {
     bool set_destination_posvel(const Vector3f& destination, const Vector3f& velocity, bool use_yaw = false, float yaw_cd = 0.0, bool use_yaw_rate = false, float yaw_rate_cds = 0.0, bool yaw_relative = false);
+    bool set_destination_posvelaccel(const Vector3f& destination, const Vector3f& velocity, const Vector3f& acceleration, bool use_yaw = false, float yaw_cd = 0.0, bool use_yaw_rate = false, float yaw_rate_cds = 0.0, bool yaw_relative = false);
+
+    Vector3f& get_target_pos() const;
```
we should put a const in front of the "Vector3f&" or the caller could change it externally I think.

rmackay9 commented on ArduCopter/mode.h:
```diff
@@ -886,19 +895,21 @@ class ModeGuided : public Mode {
-        IgnorePilotYaw    = (1U << 2),
+        IgnorePilotYaw      = (1U << 2),
         SetAttitudeTarget_ThrustAsThrust = (1U << 3),
+        DoNotStabilizePositionXY = (1U << 4),
```
I think this means we want users to be able to turn off the position and/or velocity control in Guided mode.  If "yes" then we should modify the GUID_OPTIONS parameter description to make the checkboxes visible to users.

rmackay9 commented on ArduCopter/GCS_Mavlink.cpp:
```diff
@@ -1125,7 +1125,7 @@ void GCS_MAVLINK_Copter::handleMessage(const mavlink_message_t &msg)
         bool yaw_rate_ignore = packet.type_mask & MAVLINK_SET_POS_TYPE_MASK_YAW_RATE_IGNORE;
 
         // exit immediately if acceleration provided
```
we should update this comment to, "// exit immediately if neither position nor velocity is provided"

rmackay9 commented on ArduCopter/mode.h:
```diff
@@ -845,11 +845,20 @@ class ModeGuided : public Mode {
     bool set_destination(const Vector3f& destination, bool use_yaw = false, float yaw_cd = 0.0, bool use_yaw_rate = false, float yaw_rate_cds = 0.0, bool yaw_relative = false, bool terrain_alt = false);
     bool set_destination(const Location& dest_loc, bool use_yaw = false, float yaw_cd = 0.0, bool use_yaw_rate = false, float yaw_rate_cds = 0.0, bool yaw_relative = false);
     bool get_wp(Location &loc) override;
+    void set_accel(const Vector3f& acceleration, bool use_yaw = false, float yaw_cd = 0.0, bool use_yaw_rate = false, float yaw_rate_cds = 0.0, bool yaw_relative = false, bool log_request = true);
```
I think maybe, "set_accel" isn't implemented so we can remove this line?

lthall replied:
It probably should be.....

rmackay9 commented on ArduCopter/mode.h:
```diff
@@ -845,11 +845,20 @@ class ModeGuided : public Mode {
 
     // returns true if GUIDED_OPTIONS param suggests SET_ATTITUDE_TARGET's "thrust" field should be interpreted as thrust instead of climb rate
     bool set_attitude_target_provides_thrust() const;
+    bool stop_stabilize_pos_xy() const;
```
we should put a comment above these "stop_stabilize_xx" methods.  I think also the name might be improved by changing it to something like "stabilizing_pos_xy()" and "stabilizing_vel_xy()".  As it is the name could be misinterpreted as meaning "I want to stop stabilizeing position xy" when really it's just a getter.  It would probably also be good to add WARN_IF_UNUSED at the end to avoid misuse.

rmackay9 commented on ArduCopter/mode.h:
```diff
@@ -886,19 +895,21 @@ class ModeGuided : public Mode {
     };
 
     void pos_control_start();
     void vel_control_start();
```
to keep the _start and _run methods named similarly perhaps "vel_control_start()" should be renamed to "velaccel_control_start()"

rmackay9 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -306,6 +316,13 @@ bool ModeGuided::set_destination(const Location& dest_loc, bool use_yaw, float y
 void ModeGuided::set_velocity(const Vector3f& velocity, bool use_yaw, float yaw_cd, bool use_yaw_rate, float yaw_rate_cds, bool relative_yaw, bool log_request)
+{
+    Vector3f acceleration;
+    set_velaccel(velocity, acceleration, use_yaw, yaw_cd, use_yaw_rate, yaw_rate_cds, relative_yaw, log_request);
```
It might be just as clear to pass in Vector3f() as the 2nd argument.

lthall replied:
I didn't think that worked but it did this time. I must be missing something....

rmackay9 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -316,17 +333,27 @@ void ModeGuided::set_velocity(const Vector3f& velocity, bool use_yaw, float yaw_
 bool ModeGuided::set_destination_posvel(const Vector3f& destination, const Vector3f& velocity, bool use_yaw, float yaw_cd, bool use_yaw_rate, float yaw_rate_cds, bool relative_yaw)
+{
+    Vector3f acceleration;
+    set_destination_posvelaccel(destination, velocity, acceleration, use_yaw, yaw_cd, use_yaw_rate, yaw_rate_cds, relative_yaw);
```
here too we could just set the 3rd argument as Vector3f().

rmackay9 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -13,9 +13,10 @@
+static Vector3f guided_vel_target_cms;      // velocity target (used by pos_vel_accel controller and vel_accel controller)
+static Vector3f guided_accel_target_cmss;   // acceleration target (used by pos_vel_accel controller vel_accel controller and accel controller)
+static uint32_t posvel_update_time_ms;      // system time of last target update to pos_vel_accel, vel_accel and accel controller (i.e. position and velocity update)
 static uint32_t vel_update_time_ms;         // system time of last target update to velocity controller
```
maybe we should rename:

- vel_update_time_ms to vel**accel**_update_time_ms
- posvel_update_time_ms to posvel**accel**_update_time_ms

rmackay9 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -43,7 +44,12 @@ struct Guided_Limit {
 {
     // start in position control mode
-    pos_control_start();
+    vel_control_start();
```
This init code looks a bit odd. The velocity controller is started but then it sets position targets and the posvel_update_time_ms.  My guess is that it would be easier and perhaps safer to call set_destination_posvel or set_destination_posvelaccel and provide the stopping point.

lthall replied:
Removed setting the position. Let the velocity controller bring the aircraft to a stop. That way we can't get overshoot or bounce back.

rmackay9 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -436,29 +475,44 @@ void ModeGuided::pos_control_run()
+        pos_control->input_vel_accel_z(guided_vel_target_cms, guided_accel_target_cmss, false);
+    } else {
+        guided_vel_target_cms.zero();
+        pos_control->input_pos_vel_accel_xyz(guided_pos_target_cm);
```
it's existing code but I wonder why AC_PosControl::input_pos_vel_accel_xyz is not called AC_PosControl::input_pos_xyz.

rmackay9 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -436,29 +475,44 @@ void ModeGuided::pos_control_run()
 }
 
-// guided_vel_control_run - runs the guided velocity controller
+// guided_velaccel_control_run - runs the guided velocity controller
```
partially an existing issue but we should drop the "guided_" from the start of this comment.

rmackay9 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -247,11 +253,14 @@ bool ModeGuided::set_destination(const Vector3f& destination, bool use_yaw, floa
 
-    // no need to check return status because terrain data is not used
-    wp_nav->set_wp_destination(destination, terrain_alt);
+    // record velocity target
```
I guess this comment is incorrect because it's not recording a velocity target but instead a position target.

rmackay9 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -393,7 +431,7 @@ void ModeGuided::set_angle(const Quaternion &q, float climb_rate_cms_or_thrust,
     copter.Log_Write_GuidedTarget(guided_mode,
                            Vector3f(guided_angle_state.roll_cd, guided_angle_state.pitch_cd, guided_angle_state.yaw_cd),
-                           Vector3f(0.0f, 0.0f, climb_rate_cms_or_thrust));
+                           Vector3f(0.0f, 0.0f, climb_rate_cms_or_thrust), Vector3f(0.0f, 0.0f, 0.0f));
```
"Vector3f(0.0f, 0.0f, 0.0f)" can be simplified to just, "Vector3f()" I think

rmackay9 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -408,8 +446,8 @@ void ModeGuided::takeoff_run()
         // switch to position control mode but maintain current target
-        const Vector3f target = wp_nav->get_wp_destination();
-        set_destination(target, false, 0, false, 0, false, wp_nav->origin_and_destination_are_terrain_alt());
+        const Vector3f target = pos_control->get_pos_target_cm();
```
This removes the ability to use the rangefinder as part of the takeoff in Guided mode which undoes the work of PR https://github.com/ArduPilot/ardupilot/pull/13768.  I think we need to think about this and how to maintain this existing feature.

rmackay9 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -436,29 +475,44 @@ void ModeGuided::pos_control_run()
+
+    // send position and velocity targets to position controller
+    guided_accel_target_cmss.zero();
+    if (do_avoid) {
```
I wonder what would happen if the vehicle was flying quickly to the west and the user then input a position target to the east.  Directly west of the vehicle was a fence and BendyRuler was active.  I think what would happen is the position controller init would set the target velocity to the vehicle's initial velocity.  Then this would be passed into avoid.adjust_velocity() (from within pos_control_run()).  Avoidance would shorten the desired velocity so that the vehicle doesn't hit the wall .. i guess it would be shortened and shortened and hopefully would eventually get to zero and "do_avoid" would become false and then the position controller would finally get a chance to send the vehicle in the correct direction.... so I guess it should work but I just wanted to point out that the velocity target might not have come from the current position.  I suppose a similar situation could happen if the external caller updates the position target.

lthall replied:
Removed it.

rmackay9 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -493,15 +547,34 @@ void ModeGuided::vel_control_run()
+    // update position controller with new target
+    if (stop_stabilize_vel_xy()) {
+        // set the current commanded xy vel to the desired vel
+        guided_vel_target_cms.x = pos_control->get_vel_desired_cms().x;
```
I guess this means if we're not stabilizing_vel that object avoidance doesn't work because it's overwriting the guided_vel_target_cms.x/y values that were just updated by avoid.

rmackay9 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -493,15 +547,34 @@ void ModeGuided::vel_control_run()
+        // set the current commanded xy vel to the desired vel
+        guided_vel_target_cms.x = pos_control->get_vel_desired_cms().x;
+        guided_vel_target_cms.y = pos_control->get_vel_desired_cms().y;
+        pos_control->input_vel_accel_xy(guided_vel_target_cms, guided_accel_target_cmss);
```
With a bit of re-organisation of these lines we can reduce the number of times this line appears:

> pos_control->input_vel_accel_xy(guided_vel_target_cms, guided_accel_target_cmss);

rmackay9 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -548,17 +621,48 @@ void ModeGuided::posvel_control_run()
+        guided_accel_target_cmss.zero();
+        pos_control->input_vel_accel_xy(guided_vel_target_cms, guided_accel_target_cmss);
+        pos_control->input_vel_accel_z(guided_vel_target_cms, guided_accel_target_cmss, false);
+        guided_pos_target_cm = pos_control->get_pos_target_cm();
```
I think this lines means that the position target will be lost if avoidance becomes active.  unlike velocity control modes, the position only needs to be sent in by the external caller once.. there's no time out and so no need for the caller to repeatedly pass in the target position.

lthall replied:
Removed additional functionality

rmackay9 commented:
I've gone through this in a fair bit of detail now and uncovered some issues.

I notice that this doesn't include an acceleration only control?  Acceleration is accepted but only if a velocity is also provided.  Maybe this is fine but I just want to check.

lthall commented on ArduCopter/mode_guided.cpp:
```diff
@@ -493,15 +547,34 @@ void ModeGuided::vel_control_run()
+#endif
+
+    // update position controller with new target
+    if (stop_stabilize_vel_xy()) {
```
I think I need an accel only input...

lthall replied:
The problem here is we may want to provide velocity control in Z but Accel only in X and Y.

lthall commented on ArduCopter/mode_guided.cpp:
```diff
@@ -661,17 +735,6 @@ void ModeGuided::angle_control_run()
-    pos_control->input_vel_accel_z(curr_vel_des, Vector3f(), false);
 }
 
 // helper function to set yaw state and targets
```
I think we need to add support for both Yaw and Yaw Rate.

amilcarlucas commented on ArduCopter/GCS_Mavlink.cpp:
```diff
@@ -143,20 +144,20 @@ void GCS_MAVLINK_Copter::send_position_target_local_ned()
                     POSITION_TARGET_TYPEMASK_AX_IGNORE | POSITION_TARGET_TYPEMASK_AY_IGNORE | POSITION_TARGET_TYPEMASK_AZ_IGNORE |
                     POSITION_TARGET_TYPEMASK_FORCE_SET | POSITION_TARGET_TYPEMASK_YAW_IGNORE| POSITION_TARGET_TYPEMASK_YAW_RATE_IGNORE; // ignore everything except velocity
-        target_vel = copter.flightmode->get_vel_desired_cms() * 0.01f; // convert to m/s
+        target_vel = copter.mode_guided.get_target_vel() * 0.01f; // convert to metres
```
```suggestion
        target_vel = copter.mode_guided.get_target_vel() * 0.01f; // convert to metres/s
```

amilcarlucas commented on ArduCopter/GCS_Mavlink.cpp:
```diff
@@ -143,20 +144,20 @@ void GCS_MAVLINK_Copter::send_position_target_local_ned()
-        target_pos = copter.wp_nav->get_wp_destination() * 0.01f;
-        target_vel = copter.flightmode->get_vel_desired_cms() * 0.01f;
+        target_pos = copter.mode_guided.get_target_pos() * 0.01f; // convert to metres
+        target_vel = copter.mode_guided.get_target_vel() * 0.01f; // convert to metres
```
```suggestion
        target_vel = copter.mode_guided.get_target_vel() * 0.01f; // convert to metres/s
```

lthall commented:
> I've gone through this in a fair bit of detail now and uncovered some issues.
> 
> I notice that this doesn't include an acceleration only control? Acceleration is accepted but only if a velocity is also provided. Maybe this is fine but I just want to check.

I have added this.

rmackay9 commented on ArduCopter/GCS_Mavlink.cpp:
```diff
@@ -139,20 +140,26 @@ void GCS_MAVLINK_Copter::send_position_target_local_ned()
+        target_pos = copter.mode_guided.get_target_pos() * 0.01f; // convert to metres
         break;
-    case ModeGuided::SubMode::Velocity:
+    case ModeGuided::SubMode::Accel:
```
how about we keep the order here the same as the enum?  If that means changing the order of the enum so that "Accel" is before "VelAccel" then that is OK i think.  the order of the enum will only show up to users in the form of the logging.

lthall replied:
enum class SubMode {
        TakeOff,
        WP,
        PosVelAccel,
        VelAccel,
        Accel,
        Angle,
    };
    
    take off then max autonomy to minimum autonomy.

rmackay9 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -39,16 +40,19 @@ struct Guided_Limit {
 {
     // start in position control mode
-    pos_control_start();
+    velaccel_control_start();
```
I wonder why we are changing the default submode that Guided starts in from position control to VelAccel.  If we really want to do this then we should change the comment above this line.

rmackay9 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -247,11 +279,14 @@ bool ModeGuided::set_destination(const Vector3f& destination, bool use_yaw, floa
 
-    // no need to check return status because terrain data is not used
-    wp_nav->set_wp_destination(destination, terrain_alt);
+    // set position target and zero velocity and acceleration
```
We've lost support for terrain following while in Guided mode with this change.

rmackay9 replied:
support for terrain following restored so marking this comment as resolved.

rmackay9 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -39,16 +40,19 @@ struct Guided_Limit {
     // start in position control mode
-    pos_control_start();
+    velaccel_control_start();
+    posvelaccel_update_time_ms = millis();
```
this "posvelaccel_update_time_ms" is used to ensure that we've received a position-velocity-accel target from the user within 3 seconds.  I don't think it is necessary to intiialise it here.

lthall replied:
true, removed.

rmackay9 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -286,47 +321,86 @@ bool ModeGuided::set_destination(const Location& dest_loc, bool use_yaw, float y
+    }
+    guided_vel_target_cms.zero();
+    guided_accel_target_cmss.zero();
+    vel_update_time_ms = millis();
```
I don't think we should be setting this vel_update_time_ms timer.  This is for detecting the case where the user or companion computer hasn't sent in a velocity target (and maybe it is also being re-used to detect if the user hasn't sent in an acceleration target).  In either case, I don't think we should be resetting it in the case where we've received a position target and the vehicle is happy to proceed all the way to the target (even if it is far away) without any further updates from the user.

lthall replied:
replaced both with update_time_ms

lthall replied:
Yep, removed

rmackay9 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -286,47 +321,86 @@ bool ModeGuided::set_destination(const Location& dest_loc, bool use_yaw, float y
+    guided_pos_target_cm.zero();
+    guided_vel_target_cms.zero();
+    guided_accel_target_cmss = acceleration;
+    vel_update_time_ms = millis();
```
we seem to have two timers (posvelaccel_update_time_ms and vel_update_time_ms) but I think we should have either one or three.

If we have only one, it should probably be called, "posvelaccel_update_time_ms".  If we have three they should probably be "posvelaccel_update_time_ms", "velaccel_update_time_ms" and "accel_update_time_ms".

lthall replied:
update_time_ms only

rmackay9 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -436,29 +522,31 @@ void ModeGuided::pos_control_run()
-    // run waypoint controller
-    copter.failsafe_terrain_set_status(wp_nav->update_wpnav());
+    // send position and velocity targets to position controller
+    guided_accel_target_cmss.zero();
```
I think we can safely remove these two lines to init the accel and vel targets.  I can understand the desire to be safe but there are lots of variables floating around that initialising them when not needed is a waste of flash and makes the code more complex.

lthall replied:
I sort of disagree there. This makes the code internally consistent with the position controller.

rmackay9 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -493,15 +581,107 @@ void ModeGuided::vel_control_run()
+
+    // call attitude controller
+    if (auto_yaw.mode() == AUTO_YAW_HOLD) {
+        // roll & pitch from waypoint controller, yaw rate from pilot
```
comment should probably be, "// roll & pitch from position controller, yaw rate from pilot"

rmackay9 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -493,15 +581,107 @@ void ModeGuided::vel_control_run()
     }
+}
+
+// velaccel_control_run - runs the guided velocity controller
```
comment should probably be, "runs the guided velocity and acceleration controller"

rmackay9 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -493,15 +581,107 @@ void ModeGuided::vel_control_run()
+
+// velaccel_control_run - runs the guided velocity controller
+// called from guided_run
+void ModeGuided::velaccel_control_run()
```
Not a blocker but there is certainly room for a lot of re-use of the code between accel_control_run() and velaccel_control_run().  They're almost identical except for perhaps 10 lines.

lthall replied:
yes I agree. lets get it right first.

rmackay9 commented on Tools/autotest/arducopter.py:
```diff
@@ -3639,8 +3639,8 @@ def test_guided_local_velocity_target(self, vx, vy, vz_up, timeout=3):
         # Check the last received message
-        if not (m.type_mask == 0xFFC7 or m.type_mask == 0x0FC7):
-            raise NotAchievedException("Did not receive proper mask: expected=65479 or 4039, got=%u" % m.type_mask)
+        if not (m.type_mask == 0xFE07 or m.type_mask == 0x0E07):
```
I'm surprised this change to the autotests is required..

lthall replied:
We are now reporting Acceleration too so this changed the masks.

rmackay9 commented on libraries/AP_Math/control.cpp:
```diff
@@ -210,6 +210,17 @@ void shape_accel_xy(const Vector2f& accel_input, Vector2f& accel,
     }
 }
 
+void shape_accel_xy(const Vector3f& accel_input, Vector3f& accel,
```
we should probably have a little comment above this function like we have for the others so it doesn't feel left out..

rmackay9 replied:
Ah, actually it's ok.. the function above it is the same and has lots of comments. txs

lthall replied:
It is left out, sometimes honesty is the best policy.

peterbarker reviewed (changes requested):
We really should have a test for entering guided mode from auto mode (there's a waypoint item for it).  This might break that....

tridge commented:
rebased on #17803 (done with @lthall )

tridge commented:
@lthall I've rebased -please check my fix to the tc function in AP_WPNav

lthall commented:
@tridge Looks good mate.

tridge commented:
@lthall we flew this PR yesterday on a 250 quad:
http://uav.tridgell.net/tmp/00000179.BIN
only issue was some toilet bowling in GUIDED mode

rmackay9 commented on libraries/AC_AttitudeControl/AC_PosControl.cpp:
```diff
@@ -328,15 +328,18 @@ AC_PosControl::AC_PosControl(AP_AHRS_View& ahrs, const AP_InertialNav& inav,
 ///     The time constant also defines the time taken to achieve the maximum acceleration.
 ///     The function alters the input velocity to be the velocity that the system could reach zero acceleration in the minimum time.
-void AC_PosControl::input_pos_vel_accel_xyz(const Vector3p& pos)
+void AC_PosControl::input_pos_xyz(const Vector3p& pos, const float pos_offset_z)
```
let's remove the "const" berfore "float pos_offset_z"

rmackay9 commented on libraries/AC_AttitudeControl/AC_PosControl.cpp:
```diff
@@ -815,6 +847,21 @@ void AC_PosControl::set_alt_target_with_slew(const float& pos)
 }
 
+/// update_pos_offset_z - updates the vertical offsets used by terrain following
+void AC_PosControl::update_pos_offset_z(const float pos_offset_z)
```
let's remove the "const" before "float pos_offset_z" here too (and anywhere else we find it).  The const really doesn't add any value I think and it makes the argument list longer than it needs to be.

rmackay9 commented on libraries/AC_AttitudeControl/AC_PosControl.h:
```diff
@@ -406,6 +422,9 @@ class AC_PosControl
     Vector3f    _accel_target;          // acceleration target in NEU cm/s/s
     Vector3f    _limit_vector;          // the direction that the position controller is limited, zero when not limited
     Vector2f    _vehicle_horiz_vel;     // velocity to use if _flags.vehicle_horiz_vel_override is set
+    float       _pos_offset_z;            // vertical position offset in NEU cm from home
```
nitpick: there's a small indenting problem in the comments on these three new lines.

rmackay9 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -286,47 +315,89 @@ bool ModeGuided::set_destination(const Location& dest_loc, bool use_yaw, float y
 }
 
-// guided_set_velocity - sets guided mode's target velocity
+// set_velaccel - sets guided mode's target velocity and acceleration
```
nitpick: tiny comment issue in that the comments don't match the method.  I guess this method only sets the accel targets.

rmackay9 commented:
Some issues discovered during testing this PR in Guided mode:

- [ ] takeoff when using rangefinder has a little bump that it didn't have before
![guided-takeoff-terr-before-after](https://user-images.githubusercontent.com/1498098/124701730-09256c00-df2a-11eb-936b-51f97423b30d.png).  ("param load ../Tools/autotest/default_params/copter-rangefinder.parm", "arm throttle", "takeoff 20")
- [ ] when using position control (e.g. click on map and select "Fly Here') the target position is always to home (e.g. the green line extends from the vehicle to home).
![lthall-guided-pos-targets-remains-at-home](https://user-images.githubusercontent.com/1498098/124701843-37a34700-df2a-11eb-9db7-c43d1916fbea.png)
- [ ] when using position control with a terrain altitude, the CTUN.DAlt seems to be an alt-above-terrain instead of an alt-above-origin.
- [ ] Guided position control with a terrain altitude does not maintain the correct altitude above terrain in both master and this PR.  It seems that perhaps the terrain offset is backwards but this is just a guess.  This can be reproduced (I think) using SITL with "sim_vehicle.py -L Karuizawa" and giving a Guided mode position target of, "SET_POSITION_TARGET_GLOBAL_INT 0 1 0 11 3576 363215670  1386596701 20 0 0 0 0 0 0 0 0"
- [ ] WPNAV_SPEED limit does not take effect if the parameter is changed while the vehicle is in Guided mode even if the vehicle is switched out and then back into Guided mode.  The issue can be reproduced like this:
    - take off in Loiter
    - param set WPNAV_SPEED 1000 (to set speed to 10m/s)
    - velocity 20 0 0 0 (to fly forward at 20m/s)
    - vehicle will fly at 10m/s (correct)
    - param set WPNAV_SPEED 500 (to set speed limit to 5m/s)
    - velocity 20 0 0 0 (to fly forward at 20m/s)
    - vehicle will still fly forward at 10m/s (incorrect but acceptable if we tell users that the param change will only take effect when the leave/re-enter Guided)
    - loiter (to change back to Loiter)
    - Guided (to change back to Guided mode)
    - velocity 20 0 0 0 (to fly forward at 20m/s)
    - the vehicle will still fly forward at 10m/s  <-- this is the odd thing

I still have some more things to test:

- [ ] attitude control
- [ ] acceleration control
- [ ] yaw control is OK (one issue discovered but exists in master too https://github.com/ArduPilot/ardupilot/issues/17951)
- [ ] switching to guided mode while moving
- [ ] rapidly updating position targets

rmackay9 commented on ArduCopter/GCS_Mavlink.cpp:
```diff
@@ -1169,11 +1187,13 @@ void GCS_MAVLINK_Copter::handleMessage(const mavlink_message_t &msg)
         if (!pos_ignore && !vel_ignore) {
-            copter.mode_guided.set_destination_posvel(pos_vector, vel_vector, !yaw_ignore, yaw_cd, !yaw_rate_ignore, yaw_rate_cds, yaw_relative);
+            copter.mode_guided.set_destination_posvelaccel(pos_vector, vel_vector, accel_vector, !yaw_ignore, yaw_cd, !yaw_rate_ignore, yaw_rate_cds, yaw_relative);
+        } else if (pos_ignore && vel_ignore && !acc_ignore) {
```
I think we are missing this code in the MAVLINK_MSG_ID_SET_POSITION_TARGET_GLOBAL_INT handler a little lower down in this file.

rmackay9 replied:
actually these lines can never be called because of the line, "if (pos_ignore && vel_ignore) {" above

lthall commented:
> takeoff when using rangefinder has a little bump that it didn't have before

I think this is a combination of a more assertive take off with a poorly tuned alt hold controller.

>>  when using position control (e.g. click on map and select "Fly Here') the target position is always to home (e.g. the green line extends from the vehicle to home).

I think I fixed this.

>> Guided position control with a terrain altitude does not maintain the correct altitude above terrain in both master and this PR

Fixed I think.

>> WPNAV_SPEED limit does not take effect if the parameter is changed while the vehicle is in Guided mode even if the vehicle is switched out and then back into Guided mode.

For me it take effect when we switch into guided or if we use a different guided command. I have set up the mechanisms to implement a do change speed though.

>> I noticed that you could take off in guided posvelaccel modes with any amount of climb rate. This isn't safe and I can't see this being something we want to enable so I have removed it. This would force people to give a guided take off command if the aircraft is landed.

rmackay9 commented on ArduCopter/autoyaw.cpp:
```diff
@@ -181,10 +198,16 @@ float Mode::AutoYaw::yaw()
         // keep heading pointing in the direction held in fixed_yaw
         // with no pilot input allowed
-        return _fixed_yaw;
+        float dt = millis() - fixed_last_update;
```
We can optimise this slightly by doing this:

uint32_t now_ms = millis().

rmackay9 commented:
I have very thoroughly tested this and I'm happy for this to go in now.

rmackay9 commented:
Merged, thanks!
